File: src/1171-shortest-path-in-binary-matrix/solution.py
from typing import List
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        """
        This is essentially BFS, count the layers it takes to get to rows - 1, cols - 1 coordinates

        Time: O(N * M)
        Space: O(N * M)
        """
        directions = [(-1, -1), (1, 1), (-1, 0), (0, -1), (1, 0), (0, 1), (1, -1), (-1, 1)]
        rows, cols = len(grid), len(grid[0])

        # Edge case: We are blocked from the start and the end, we can't even move!
        if (
            grid[-1][-1] == 1 or
            grid[0][0] == 1
        ):
            return -1

        def bfs(r: int, c: int) -> int:
            length = 1
            q = deque([(r, c)])
            visited = set((r, c))

            while q:
                # Go through layer by layer in the queue
                #! NOTE: This is a modified layer by layer BFS
                #! NOTE: You usually dont need this
                for _ in range(len(q)):
                    curr_r, curr_c = q.popleft()
                    
                    # If reached destination just return length
                    if curr_r == rows - 1 and curr_c == cols - 1:
                        return length

                    # Go through each directions
                    for dr, dc in directions:
                        new_r, new_c = dr + curr_r, dc + curr_c
                        if (
                            new_r < 0 or new_r >= rows or
                            new_c < 0 or new_c >= cols or
                            (new_r, new_c) in visited or
                            grid[new_r][new_c] == 1
                        ):
                            continue
                        q.append([new_r, new_c])
                        visited.add((new_r, new_c))
                # After emptying current q, we gone through a layer
                # which means we travelled one unit forward
                length += 1

            return -1
        return bfs(0, 0)

File: src/1171-shortest-path-in-binary-matrix/solution.ts
function shortestPathBinaryMatrix(grid: number[][]): number {
    const directions = [[-1, -1], [1, 1], [-1, 0], [0, -1], [1, 0], [0, 1], [1, -1], [-1, 1]];
    const rows = grid.length;
    const cols = grid[0].length;
    const q = new Queue.Queue();
    const visited = new Set<string>(); 

    // Edge case: We are blocked from the start and the end, we can't even move!
    if (grid[rows - 1][cols - 1] === 1 || grid[0][0] === 1) {
        return -1;
    }

    // Edge case: Single cell that is open
    if (grid[0][0] === 0 && rows === 1 && cols === 1) {
        return 1;
    }

    function bfs(r: number, c: number): number {
        q.enqueue([r, c]);
        visited.add(`${r},${c}`);
        let area = 1;

        while (q.size()) {
            const levelSize = q.size();
            for (let i = 0; i < levelSize; i++) {
                const [currentRow, currentCol] = q.dequeue();

                for (const [dr, dc] of directions) {
                    const newRow = currentRow + dr;
                    const newCol = currentCol + dc;

                    // We reached the end, return the path length
                    if (newRow === rows - 1 && newCol === cols - 1) {
                        return area + 1;
                    }

                    if (
                        newRow < 0 || newRow >= rows ||
                        newCol < 0 || newCol >= cols ||
                        visited.has(`${newRow},${newCol}`) ||
                        grid[newRow][newCol] === 1
                    ) {
                        continue;
                    }
                    q.enqueue([newRow, newCol]);
                    visited.add(`${newRow},${newCol}`);
                }
            }
            area++;
        }

        // Return -1 if no path, we should find the path after traversal
        return -1;
    }

    return bfs(0, 0);
}
